File: src/client/VsClk.SignalService/Client/HubProxyBase.cs
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.SignalR.Client;

namespace Microsoft.VsCloudKernel.SignalService.Client
{
    /// <summary>
    /// Our client proxy base class
    /// </summary>
    public class HubProxyBase
    {
        private readonly string hubName;
        private const string InvokeHubMethodAsync = "InvokeHubMethodAsync";

        protected HubProxyBase(HubConnection connection, string hubName)
        {
            Connection = Requires.NotNull(connection, nameof(connection));
            this.hubName = hubName;
        }

        protected HubConnection Connection { get; }

        protected string ToHubMethodName(string methodName)
        {
            return string.IsNullOrEmpty(this.hubName) ? methodName : $"{this.hubName}.{methodName}";
        }

        protected async Task<T> InvokeAsync<T>(string methodName, object[] args, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (string.IsNullOrEmpty(this.hubName))
            {
                return (T)(await Connection.InvokeCoreAsync(methodName, typeof(T), args, cancellationToken).ConfigureAwait(false));
            }
            else
            {
                return (T)(await Connection.InvokeCoreAsync(InvokeHubMethodAsync, typeof(T), new object[] { ToHubMethodName(methodName), args }, cancellationToken));
            }
        }

        protected Task InvokeAsync(string methodName, object[] args, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (string.IsNullOrEmpty(this.hubName))
            {
                return Connection.InvokeCoreAsync(methodName, typeof(object), args, cancellationToken);
            }
            else
            {
                return Connection.InvokeCoreAsync(InvokeHubMethodAsync, typeof(object), new object[] { ToHubMethodName(methodName), args }, cancellationToken);
            }
        }
    }
}

File: src/client/VsClk.SignalService/ts/src/HubProxyBase.ts
import * as signalR from '@microsoft/signalr';

export class HubProxyBase {
    private static readonly invokeHubMethodAsync = 'InvokeHubMethodAsync';

    constructor(
        public readonly hubConnection: signalR.HubConnection,
        protected readonly logger?: signalR.ILogger,
        protected readonly hubName?: string) {
    }

    protected toHubMethodName(methodName: string): string {
        return this.hubName ? `${this.hubName}.${methodName}` : methodName;
    }

    public invoke<T = any>(methodName: string, ...args: any[]): Promise<T> {
        if (this.hubName) {
            return this.hubConnection.invoke<T>(HubProxyBase.invokeHubMethodAsync, this.toHubMethodName(methodName), args );
        } else {
            return this.hubConnection.invoke<T>(methodName, ...args);
        }
    }

    public send(methodName: string, ...args: any[]): Promise<void> {
        if (this.hubName) {
            return this.hubConnection.send(HubProxyBase.invokeHubMethodAsync, this.toHubMethodName(methodName), args );
        } else {
            return this.hubConnection.send(methodName, ...args);
        }     
    }
}